File: NoTomorrow/Features/Progress/ProgressPhotosViews.swift
import SwiftUI
import SwiftData
import PhotosUI

/// A stored progress photo at a display size; a surface placeholder (or nothing) while it decodes.
struct ProgressPhotoImage: View {
    var url: URL
    var maxPixel: CGFloat
    /// Fill (and crop) the frame, or fit inside it.
    var fill: Bool = true

    @State private var image: UIImage?

    var body: some View {
        ZStack {
            if let image {
                Image(uiImage: image)
                    .resizable()
                    .aspectRatio(contentMode: fill ? .fill : .fit)
            } else {
                Rectangle().fill(fill ? NT.Colors.surface2 : Color.clear)
            }
        }
        .task(id: url) {
            image = await ProgressPhotoLoader.image(at: url, maxPixel: maxPixel)
        }
    }
}

/// Progress > Body, under the measurements: private progress photos. A strip of thumbnails (newest first) after
/// an Add photo tile, a full-screen viewer and Compare (two photos side by side). Photos never leave the phone.
struct ProgressPhotosSection: View {
    private struct ViewerStart: Identifiable {
        let id: UUID
    }

    private static let tileWidth: CGFloat = 84
    private static let tileHeight: CGFloat = 112

    @Environment(\.modelContext) private var modelContext
    @Query(sort: \ProgressPhoto.takenAt, order: .reverse) private var rows: [ProgressPhoto]
    @State private var pickerItem: PhotosPickerItem?
    @State private var pendingJPEG: Data?
    @State private var asksPose = false
    @State private var isImporting = false
    @State private var showsFailure = false
    @State private var viewerStart: ViewerStart?
    @State private var showsCompare = false

    private let store = ProgressPhotoStore.standard

    private var photos: [ProgressPhoto] {
        ProgressPhotos.newestFirst(rows, takenAt: \.takenAt, id: \.id)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            header
            Text("photos.private")
                .font(NT.Fonts.footnote).foregroundStyle(NT.Colors.ink2)
                .fixedSize(horizontal: false, vertical: true)
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 10) {
                    addTile
                    ForEach(photos) { photo in
                        thumbnail(photo)
                    }
                }
            }
            .padding(.top, 12)
        }
        .onChange(of: pickerItem) { _, item in
            guard let item else { return }
            isImporting = true
            Task {
                let data = try? await item.loadTransferable(type: Data.self)
                var jpeg: Data?
                if let data {
                    jpeg = await PickedPhoto.data(data).preparedJPEG(maxLongEdge: ProgressPhotos.maxLongEdge)
                }
                let prepared = jpeg
                await MainActor.run {
                    pickerItem = nil
                    isImporting = false
                    if let prepared {
                        pendingJPEG = prepared
                        asksPose = true
                    } else {
                        showsFailure = true
                    }
                }
            }
        }
        .confirmationDialog("photos.pose.question", isPresented: $asksPose, titleVisibility: .visible) {
            ForEach(ProgressPose.allCases, id: \.self) { pose in
                Button(ProgressPhotos.poseKey(pose)) { save(pose: pose) }
            }
            Button("photos.pose.none") { save(pose: nil) }
            Button("common.cancel", role: .cancel) { pendingJPEG = nil }
        }
        .alert("photos.failed", isPresented: $showsFailure) {
            Button("common.done", role: .cancel) {}
        }
        .fullScreenCover(item: $viewerStart) { start in
            ProgressPhotoViewer(startID: start.id, store: store)
        }
        .fullScreenCover(isPresented: $showsCompare) {
            ProgressPhotoCompareView(store: store)
        }
    }

    private var header: some View {
        HStack {
            Text("photos.title").font(NT.Fonts.headline).foregroundStyle(NT.Colors.ink)
            Spacer()
            if photos.count >= 2 {
                Button { showsCompare = true } label: {
                    HStack(spacing: 4) {
                        Image(systemName: "rectangle.split.2x1").font(.system(size: 14, weight: .semibold))
                            .accessibilityHidden(true)
                        Text("photos.compare").font(NT.Fonts.subheadlineBold)
                    }
                    .foregroundStyle(NT.Colors.ember)
                    .frame(minHeight: NT.Size.control)
                }
                .buttonStyle(PressScale())
            }
        }
        .frame(minHeight: NT.Size.control)
    }

    private var addTile: some View {
        PhotosPicker(selection: $pickerItem, matching: .images, photoLibrary: .shared()) {
            VStack(spacing: 6) {
                if isImporting {
                    ProgressView().tint(NT.Colors.ink)
                } else {
                    Image(systemName: "plus").font(.system(size: 20, weight: .semibold))
                        .accessibilityHidden(true)
                }
                Text("photos.add")
                    .font(NT.Fonts.caption)
                    .multilineTextAlignment(.center)
                    .lineLimit(2)
            }
            .foregroundStyle(NT.Colors.ink)
            .padding(.horizontal, 6)
            .frame(width: Self.tileWidth, height: Self.tileHeight)
            .background(NT.Colors.surface2, in: RoundedRectangle(cornerRadius: NT.Radius.cell, style: .continuous))
        }
        .buttonStyle(PressScale())
        .disabled(isImporting)
    }

    private func thumbnail(_ photo: ProgressPhoto) -> some View {
        Button { viewerStart = ViewerStart(id: photo.id) } label: {
            ProgressPhotoImage(url: store.url(for: photo.fileName), maxPixel: 320)
                .frame(width: Self.tileWidth, height: Self.tileHeight)
                .clipShape(RoundedRectangle(cornerRadius: NT.Radius.cell, style: .continuous))
                .contentShape(RoundedRectangle(cornerRadius: NT.Radius.cell, style: .continuous))
                .overlay(alignment: .bottomLeading) {
                    Text(verbatim: Fmt.dayMonth(photo.takenAt))
                        .font(NT.Fonts.caption)
                        .foregroundStyle(Color.white)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(Color.black.opacity(0.55), in: Capsule())
                        .padding(5)
                }
        }
        .buttonStyle(PressScale())
        .accessibilityLabel(Text(verbatim: ProgressPhotos.accessibilityLabel(takenAt: photo.takenAt, pose: photo.pose)))
    }

    private func save(pose: ProgressPose?) {
        guard let jpeg = pendingJPEG else { return }
        pendingJPEG = nil
        do {
            try ProgressPhotos.add(jpeg: jpeg, pose: pose, store: store, in: modelContext)
            Haptics.success()
        } catch {
            showsFailure = true
        }
    }
}

// MARK: - Viewer

/// Full screen: swipe between the photos (newest first), date and pose under each, Delete with a confirmation.
struct ProgressPhotoViewer: View {
    var store: ProgressPhotoStore

    @Environment(\.modelContext) private var modelContext
    @Environment(\.dismiss) private var dismiss
    @Query(sort: \ProgressPhoto.takenAt, order: .reverse) private var rows: [ProgressPhoto]
    @State private var selection: UUID?
    @State private var confirmsDelete = false

    init(startID: UUID, store: ProgressPhotoStore) {
        self.store = store
        _selection = State(initialValue: startID)
    }

    private var photos: [ProgressPhoto] {
        ProgressPhotos.newestFirst(rows, takenAt: \.takenAt, id: \.id)
    }

    private var currentIndex: Int? {
        photos.firstIndex { $0.id == selection }
    }

    var body: some View {
        let list = photos
        let index = currentIndex ?? 0
        VStack(spacing: 0) {
            HStack {
                Button { dismiss() } label: {
                    Image(systemName: "xmark")
                        .font(.system(size: 17, weight: .semibold))
                        .foregroundStyle(NT.Colors.ink)
                        .frame(width: NT.Size.control, height: NT.Size.control)
                }
                .accessibilityLabel(Text("common.done"))
                Spacer()
                if list.count > 1 {
                    Text(verbatim: "\(index + 1) / \(list.count)")
                        .font(NT.Fonts.subheadline).foregroundStyle(NT.Colors.ink2).tabular()
                }
                Spacer()
                Button { confirmsDelete = true } label: {
                    Image(systemName: "trash")
                        .font(.system(size: 17, weight: .semibold))
                        .foregroundStyle(NT.Colors.bad)
                        .frame(width: NT.Size.control, height: NT.Size.control)
                }
                .accessibilityLabel(Text("common.delete"))
                .disabled(list.isEmpty)
            }
            .padding(.horizontal, 8)

            TabView(selection: $selection) {
                ForEach(list) { photo in
                    ProgressPhotoImage(url: store.url(for: photo.fileName), maxPixel: 2048, fill: false)
                        .frame(maxWidth: .infinity, maxHeight: .infinity)
                        .accessibilityElement(children: .ignore)
                        .accessibilityLabel(Text(verbatim: ProgressPhotos.accessibilityLabel(takenAt: photo.takenAt,
                                                                                             pose: photo.pose)))
                        .accessibilityAddTraits(.isImage)
                        .tag(Optional(photo.id))
                }
            }
            .tabViewStyle(.page(indexDisplayMode: .never))

            if index < list.count {
                caption(list[index])
            }
        }
        .background(NT.Colors.ground.ignoresSafeArea())
        .confirmationDialog("photos.delete.title", isPresented: $confirmsDelete, titleVisibility: .visible) {
            Button("common.delete", role: .destructive) { deleteCurrent() }
            Button("common.cancel", role: .cancel) {}
        } message: {
            Text("photos.delete.message")
        }
    }

    private func caption(_ photo: ProgressPhoto) -> some View {
        VStack(spacing: 4) {
            Text(verbatim: ProgressPhotos.dateLabel(photo.takenAt))
                .font(NT.Fonts.headline).foregroundStyle(NT.Colors.ink)
            if let pose = photo.pose {
                Text(ProgressPhotos.poseKey(pose))
                    .font(NT.Fonts.subheadline).foregroundStyle(NT.Colors.ink2)
            }
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, 16)
    }

    private func deleteCurrent() {
        let list = photos
        guard let index = currentIndex, index < list.count else { return }
        let photo = list[index]
        let next = ProgressPhotos.indexAfterDeleting(at: index, count: list.count)
        let remaining = list.filter { $0.id != photo.id }
        ProgressPhotos.delete(photo, store: store, in: modelContext)
        if let next, next < remaining.count {
            selection = remaining[next].id
        } else {
            dismiss()
        }
    }
}

// MARK: - Compare

/// Two photos side by side: the first against the latest to start with; each side picks any photo from a menu.
struct ProgressPhotoCompareView: View {
    var store: ProgressPhotoStore

    @Environment(\.dismiss) private var dismiss
    @Query(sort: \ProgressPhoto.takenAt, order: .reverse) private var rows: [ProgressPhoto]
    @State private var beforeID: UUID?
    @State private var afterID: UUID?

    private var photos: [ProgressPhoto] {
        ProgressPhotos.newestFirst(rows, takenAt: \.takenAt, id: \.id)
    }

    var body: some View {
        let list = photos
        let defaults = ProgressPhotos.defaultComparison(list)
        VStack(spacing: 0) {
            ZStack {
                Text("photos.compare").font(NT.Fonts.headline).foregroundStyle(NT.Colors.ink)
                HStack {
                    Button { dismiss() } label: {
                        Image(systemName: "xmark")
                            .font(.system(size: 17, weight: .semibold))
                            .foregroundStyle(NT.Colors.ink)
                            .frame(width: NT.Size.control, height: NT.Size.control)
                    }
                    .accessibilityLabel(Text("common.done"))
                    Spacer()
                }
            }
            .padding(.horizontal, 8)

            if let defaults {
                HStack(alignment: .top, spacing: 8) {
                    column(list.first { $0.id == beforeID } ?? defaults.before, options: list,
                           title: "photos.before", selection: $beforeID)
                    column(list.first { $0.id == afterID } ?? defaults.after, options: list,
                           title: "photos.after", selection: $afterID)
                }
                .padding(.horizontal, 12)
                .padding(.top, 16)
            }
            Spacer(minLength: 0)
        }
        .background(NT.Colors.ground.ignoresSafeArea())
    }

    private func column(_ photo: ProgressPhoto, options: [ProgressPhoto], title: LocalizedStringKey,
                        selection: Binding<UUID?>) -> some View {
        VStack(spacing: 8) {
            Color.clear
                .aspectRatio(3.0 / 4.0, contentMode: .fit)
                .overlay {
                    ProgressPhotoImage(url: store.url(for: photo.fileName), maxPixel: 1200)
                }
                .clipShape(RoundedRectangle(cornerRadius: NT.Radius.cell, style: .continuous))
                .accessibilityElement(children: .ignore)
                .accessibilityLabel(Text(title) + Text(verbatim: ": "
                    + ProgressPhotos.accessibilityLabel(takenAt: photo.takenAt, pose: photo.pose)))
                .accessibilityAddTraits(.isImage)
            Menu {
                ForEach(options) { option in
                    Button {
                        selection.wrappedValue = option.id
                    } label: {
                        Text(verbatim: ProgressPhotos.optionLabel(takenAt: option.takenAt, pose: option.pose))
                    }
                }
            } label: {
                VStack(spacing: 2) {
                    Text(title).font(NT.Fonts.eyebrow).foregroundStyle(NT.Colors.ink2)
                    HStack(spacing: 4) {
                        Text(verbatim: ProgressPhotos.optionLabel(takenAt: photo.takenAt, pose: photo.pose))
                            .font(NT.Fonts.footnoteBold)
                            .lineLimit(1)
                            .minimumScaleFactor(0.8)
                        Image(systemName: "chevron.down").font(.system(size: 11, weight: .semibold))
                            .accessibilityHidden(true)
                    }
                    .foregroundStyle(NT.Colors.ink)
                }
                .frame(maxWidth: .infinity, minHeight: NT.Size.control)
            }
        }
        .frame(maxWidth: .infinity)
    }
}

File: android/app/src/main/java/app/notomorrow/feature/progress/ProgressPhotosViews.kt
package app.notomorrow.feature.progress

import android.graphics.BitmapFactory
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.systemBarsPadding
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.foundation.rememberScrollState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import app.notomorrow.data.entity.ProgressPhotoEntity
import app.notomorrow.data.files.ProgressPhotoFiles
import app.notomorrow.designsystem.Eyebrow
import app.notomorrow.designsystem.NT
import app.notomorrow.designsystem.NtActionSheet
import app.notomorrow.designsystem.NtAlert
import app.notomorrow.designsystem.NtAlertAction
import app.notomorrow.designsystem.NtAlertRole
import app.notomorrow.designsystem.NtFullScreenCover
import app.notomorrow.designsystem.NtIcon
import app.notomorrow.designsystem.NtIcons
import app.notomorrow.designsystem.NtMenu
import app.notomorrow.designsystem.NtMenuItem
import app.notomorrow.designsystem.NtShapes
import app.notomorrow.designsystem.NtSpinner
import app.notomorrow.designsystem.NtText
import app.notomorrow.designsystem.appLocale
import app.notomorrow.designsystem.ntPlainClickable
import app.notomorrow.designsystem.pressScale
import app.notomorrow.di.LocalAppContainer
import app.notomorrow.model.ProgressPose
import app.notomorrow.util.ImageDownscaler
import app.notomorrow.util.NtKeys
import app.notomorrow.util.S
import java.io.File
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

private val TileWidth = 84.dp
private val TileHeight = 112.dp

/**
 * Progress › Body, under the measurements — `ProgressPhotosSection`: private progress photos. A
 * strip of thumbnails (newest first) after an Add photo tile (the system Photo Picker, no storage
 * permission), a full-screen viewer and Compare (two photos side by side). Photos never leave the
 * phone. No camera here: iOS offers only the library too.
 */
@Composable
fun ProgressPhotosSection(modifier: Modifier = Modifier) {
    val context = LocalContext.current
    val dao = LocalAppContainer.current.db.progressPhotoDao()
    val files = remember(context) { ProgressPhotoFiles.inFilesDir(context.filesDir) }
    val rows by remember(dao) { dao.observeAll() }.collectAsState(initial = emptyList())
    val photos = ProgressPhotos.newestFirst(rows)
    val scope = rememberCoroutineScope()
    val haptics = LocalHapticFeedback.current

    var importing by remember { mutableStateOf(false) }
    var pendingJpeg by remember { mutableStateOf<ByteArray?>(null) }
    var showsFailure by remember { mutableStateOf(false) }
    var viewerStart by remember { mutableStateOf<String?>(null) }
    // Kept after close so the cover's exit slide still has something to draw.
    var lastViewerStart by remember { mutableStateOf<String?>(null) }
    var showsCompare by remember { mutableStateOf(false) }

    val picker = rememberLauncherForActivityResult(ActivityResultContracts.PickVisualMedia()) { uri ->
        if (uri != null) {
            importing = true
            scope.launch {
                val jpeg = withContext(Dispatchers.IO) {
                    runCatching {
                        ImageDownscaler.jpeg(context, uri, ProgressPhotos.MAX_LONG_EDGE, ProgressPhotos.QUALITY)
                    }.getOrNull()
                }
                importing = false
                if (jpeg != null) pendingJpeg = jpeg else showsFailure = true
            }
        }
    }

    fun save(jpeg: ByteArray, pose: ProgressPose?) {
        pendingJpeg = null
        scope.launch {
            val saved = runCatching { ProgressPhotos.add(jpeg, pose, dao, files) }
            if (saved.isSuccess) haptics.performHapticFeedback(HapticFeedbackType.Confirm) else showsFailure = true
        }
    }

    Column(modifier.fillMaxWidth()) {
        Row(
            modifier = Modifier.fillMaxWidth().heightIn(min = NT.Size.control),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            NtText(
                text = stringResource(S.photos_title),
                modifier = Modifier.weight(1f),
                style = NT.Fonts.headline,
                color = NT.Colors.ink,
                maxLines = 1,
            )
            if (photos.size >= 2) {
                Box(
                    modifier = Modifier.heightIn(min = NT.Size.control).pressScale { showsCompare = true },
                    contentAlignment = Alignment.Center,
                ) {
                    NtText(
                        stringResource(S.photos_compare),
                        style = NT.Fonts.subheadlineBold,
                        color = NT.Colors.ember,
                        maxLines = 1,
                    )
                }
            }
        }
        NtText(stringResource(S.photos_private), style = NT.Fonts.footnote, color = NT.Colors.ink2)
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 12.dp)
                .horizontalScroll(rememberScrollState()),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            AddPhotoTile(importing = importing) {
                picker.launch(PickVisualMediaRequest(ActivityResultContracts.PickVisualMedia.ImageOnly))
            }
            photos.forEach { photo ->
                PhotoThumbnail(photo, files) {
                    viewerStart = photo.id
                    lastViewerStart = photo.id
                }
            }
        }
    }

    pendingJpeg?.let { jpeg ->
        NtActionSheet(
            title = stringResource(S.photos_pose_question),
            actions = ProgressPose.entries.map { pose ->
                NtAlertAction(stringResource(NtKeys.pose(pose))) { save(jpeg, pose) }
            } + NtAlertAction(stringResource(S.photos_pose_none)) { save(jpeg, null) },
            cancel = stringResource(S.common_cancel),
            onDismiss = { pendingJpeg = null },
        )
    }

    if (showsFailure) {
        NtAlert(
            title = stringResource(S.photos_failed),
            actions = listOf(NtAlertAction(stringResource(S.common_done), NtAlertRole.Cancel)),
            onDismiss = { showsFailure = false },
        )
    }

    NtFullScreenCover(visible = viewerStart != null, onDismiss = { viewerStart = null }) {
        ProgressPhotoViewer(
            photos = photos,
            startId = lastViewerStart,
            files = files,
            onDelete = { photo -> scope.launch { ProgressPhotos.delete(photo, dao, files) } },
            onClose = { viewerStart = null },
        )
    }

    NtFullScreenCover(visible = showsCompare && photos.size >= 2, onDismiss = { showsCompare = false }) {
        ProgressPhotoCompare(photos = photos, files = files, onClose = { showsCompare = false })
    }
}

@Composable
private fun AddPhotoTile(importing: Boolean, onClick: () -> Unit) {
    Column(
        modifier = Modifier
            .size(TileWidth, TileHeight)
            .pressScale(enabled = !importing, onClick = onClick)
            .background(NT.Colors.surface2, NtShapes.cell)
            .padding(horizontal = 6.dp),
        verticalArrangement = Arrangement.spacedBy(6.dp, Alignment.CenterVertically),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        if (importing) {
            NtSpinner()
        } else {
            NtIcon(NtIcons.Plus, size = 20.dp, tint = NT.Colors.ink)
        }
        NtText(
            text = stringResource(S.photos_add),
            style = NT.Fonts.caption,
            color = NT.Colors.ink,
            maxLines = 2,
            textAlign = TextAlign.Center,
        )
    }
}

@Composable
private fun PhotoThumbnail(photo: ProgressPhotoEntity, files: ProgressPhotoFiles, onClick: () -> Unit) {
    val description = photoDescription(photo)
    Box(
        modifier = Modifier
            .size(TileWidth, TileHeight)
            .pressScale(onClick = onClick)
            .semantics { contentDescription = description }
            .clip(NtShapes.cell),
    ) {
        PhotoImage(files.file(photo.fileName), maxPixel = 320, crop = true, modifier = Modifier.fillMaxSize())
        NtText(
            text = ProgressPhotos.shortDateLabel(photo.takenAt),
            modifier = Modifier
                .clearAndSetSemantics {}
                .align(Alignment.BottomStart)
                .padding(5.dp)
                .background(Color.Black.copy(alpha = 0.55f), NtShapes.capsule)
                .padding(horizontal = 6.dp, vertical = 2.dp),
            style = NT.Fonts.caption,
            color = Color.White,
            maxLines = 1,
        )
    }
}

/** TalkBack's name for a photo: "Progress photo, front, 26 September 2026". */
@Composable
private fun photoDescription(photo: ProgressPhotoEntity): String {
    val date = ProgressPhotos.dateLabel(photo.takenAt)
    val pose = ProgressPhotos.pose(photo) ?: return stringResource(S.photos_a11y, date)
    return stringResource(S.photos_a11y_pose, stringResource(NtKeys.pose(pose)).lowercase(appLocale()), date)
}

/** A stored photo decoded at about [maxPixel] on its long edge, off the main thread. */
@Composable
private fun PhotoImage(file: File, maxPixel: Int, crop: Boolean, modifier: Modifier = Modifier) {
    val bitmap by produceState<ImageBitmap?>(initialValue = null, file, maxPixel) {
        value = withContext(Dispatchers.IO) { decodePhoto(file, maxPixel) }
    }
    Box(modifier.background(if (crop) NT.Colors.surface2 else Color.Transparent)) {
        bitmap?.let {
            Image(
                bitmap = it,
                contentDescription = null,
                modifier = Modifier.fillMaxSize(),
                contentScale = if (crop) ContentScale.Crop else ContentScale.Fit,
            )
        }
    }
}

/** Stored photos are upright already (the rotation was baked in on save), so no EXIF here. */
private fun decodePhoto(file: File, maxPixel: Int): ImageBitmap? = runCatching {
    val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    BitmapFactory.decodeFile(file.path, bounds)
    if (bounds.outWidth <= 0 || bounds.outHeight <= 0) return@runCatching null
    val target = ImageDownscaler.targetSize(bounds.outWidth, bounds.outHeight, maxPixel)
        ?: return@runCatching null
    val options = BitmapFactory.Options().apply {
        inSampleSize = ImageDownscaler.sampleSize(bounds.outWidth, bounds.outHeight, target.first, target.second)
    }
    BitmapFactory.decodeFile(file.path, options)?.asImageBitmap()
}.getOrNull()

// MARK: - Viewer

/** Full screen: swipe between the photos (newest first), date and pose under each, Delete with a confirmation. */
@Composable
private fun ProgressPhotoViewer(
    photos: List<ProgressPhotoEntity>,
    startId: String?,
    files: ProgressPhotoFiles,
    onDelete: (ProgressPhotoEntity) -> Unit,
    onClose: () -> Unit,
) {
    val initial = remember { photos.indexOfFirst { it.id == startId }.coerceAtLeast(0) }
    val pager = rememberPagerState(initialPage = initial) { photos.size }
    val scope = rememberCoroutineScope()
    var confirmsDelete by remember { mutableStateOf(false) }
    val index = pager.currentPage.coerceAtMost((photos.size - 1).coerceAtLeast(0))
    val current = photos.getOrNull(index)

    Column(Modifier.fillMaxSize().systemBarsPadding()) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(horizontal = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            ViewerIcon(NtIcons.Xmark, stringResource(S.common_done), NT.Colors.ink, onClose)
            NtText(
                text = if (photos.size > 1) "${index + 1} / ${photos.size}" else "",
                modifier = Modifier.weight(1f),
                style = NT.Fonts.subheadline,
                color = NT.Colors.ink2,
                maxLines = 1,
                textAlign = TextAlign.Center,
            )
            ViewerIcon(NtIcons.Trash, stringResource(S.common_delete), NT.Colors.bad) {
                if (current != null) confirmsDelete = true
            }
        }
        HorizontalPager(
            state = pager,
            modifier = Modifier.weight(1f).fillMaxWidth(),
            key = { page -> photos.getOrNull(page)?.id ?: page },
        ) { page ->
            photos.getOrNull(page)?.let { photo ->
                val description = photoDescription(photo)
                PhotoImage(
                    files.file(photo.fileName),
                    maxPixel = 2048,
                    crop = false,
                    modifier = Modifier.fillMaxSize().semantics {
                        contentDescription = description
                        role = Role.Image
                    },
                )
            }
        }
        if (current != null) {
            Column(
                modifier = Modifier.fillMaxWidth().padding(vertical = 16.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(4.dp),
            ) {
                NtText(ProgressPhotos.dateLabel(current.takenAt), style = NT.Fonts.headline, color = NT.Colors.ink)
                ProgressPhotos.pose(current)?.let { pose ->
                    NtText(stringResource(NtKeys.pose(pose)), style = NT.Fonts.subheadline, color = NT.Colors.ink2)
                }
            }
        }
    }

    if (confirmsDelete && current != null) {
        NtAlert(
            title = stringResource(S.photos_delete_title),
            message = stringResource(S.photos_delete_message),
            actions = listOf(
                NtAlertAction(stringResource(S.common_cancel), NtAlertRole.Cancel),
                NtAlertAction(stringResource(S.common_delete), NtAlertRole.Destructive) {
                    val next = ProgressPhotos.indexAfterDeleting(index, photos.size)
                    onDelete(current)
                    if (next == null) onClose() else scope.launch { pager.scrollToPage(next) }
                },
            ),
            onDismiss = { confirmsDelete = false },
        )
    }
}

@Composable
private fun ViewerIcon(icon: NtIcons, label: String, tint: Color, onClick: () -> Unit) {
    Box(
        modifier = Modifier.size(NT.Size.control).ntPlainClickable(onClickLabel = label, onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        NtIcon(icon, size = 20.dp, tint = tint, contentDescription = label)
    }
}

// MARK: - Compare

/** Two photos side by side: the first against the latest to start with; each side picks any photo from a menu. */
@Composable
private fun ProgressPhotoCompare(
    photos: List<ProgressPhotoEntity>,
    files: ProgressPhotoFiles,
    onClose: () -> Unit,
) {
    var beforeId by remember { mutableStateOf<String?>(null) }
    var afterId by remember { mutableStateOf<String?>(null) }
    val defaults = ProgressPhotos.defaultComparison(photos)

    Column(Modifier.fillMaxSize().systemBarsPadding()) {
        Box(Modifier.fillMaxWidth().padding(horizontal = 8.dp), contentAlignment = Alignment.Center) {
            Box(Modifier.align(Alignment.CenterStart)) {
                ViewerIcon(NtIcons.Xmark, stringResource(S.common_done), NT.Colors.ink, onClose)
            }
            NtText(stringResource(S.photos_compare), style = NT.Fonts.headline, color = NT.Colors.ink, maxLines = 1)
        }
        if (defaults != null) {
            Row(
                modifier = Modifier.fillMaxWidth().padding(horizontal = 12.dp).padding(top = 16.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                CompareColumn(
                    photo = photos.firstOrNull { it.id == beforeId } ?: defaults.first,
                    options = photos,
                    title = stringResource(S.photos_before),
                    files = files,
                    onPick = { beforeId = it },
                    modifier = Modifier.weight(1f),
                )
                CompareColumn(
                    photo = photos.firstOrNull { it.id == afterId } ?: defaults.second,
                    options = photos,
                    title = stringResource(S.photos_after),
                    files = files,
                    onPick = { afterId = it },
                    modifier = Modifier.weight(1f),
                )
            }
        }
        Spacer(Modifier.weight(1f))
    }
}

@Composable
private fun CompareColumn(
    photo: ProgressPhotoEntity,
    options: List<ProgressPhotoEntity>,
    title: String,
    files: ProgressPhotoFiles,
    onPick: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    var expanded by remember { mutableStateOf(false) }
    val poseLabels = ProgressPose.entries.associateWith { stringResource(NtKeys.pose(it)) }
    fun label(row: ProgressPhotoEntity): String {
        val pose = ProgressPhotos.pose(row)?.let { poseLabels[it] }
        val date = ProgressPhotos.dateLabel(row.takenAt)
        return if (pose == null) date else "$date · $pose"
    }

    val description = title + ": " + photoDescription(photo)
    Column(modifier, verticalArrangement = Arrangement.spacedBy(8.dp), horizontalAlignment = Alignment.CenterHorizontally) {
        PhotoImage(
            files.file(photo.fileName),
            maxPixel = 1200,
            crop = true,
            modifier = Modifier
                .fillMaxWidth()
                .aspectRatio(3f / 4f)
                .clip(NtShapes.cell)
                .semantics {
                    contentDescription = description
                    role = Role.Image
                },
        )
        Box(
            modifier = Modifier.fillMaxWidth().heightIn(min = NT.Size.control).ntPlainClickable { expanded = true },
            contentAlignment = Alignment.Center,
        ) {
            Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(2.dp)) {
                Eyebrow(title)
                Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                    NtText(label(photo), style = NT.Fonts.footnoteBold, color = NT.Colors.ink, maxLines = 1)
                    NtIcon(NtIcons.ChevronDown, size = 11.dp, tint = NT.Colors.ink)
                }
            }
            NtMenu(
                expanded = expanded,
                onDismiss = { expanded = false },
                items = options.map { option ->
                    NtMenuItem(title = label(option), onClick = {
                        onPick(option.id)
                        expanded = false
                    })
                },
            )
        }
    }
}
